packages/in_app_purchase/in_app_purchase_platform_interface/lib/src/errors/purchase_exception.dart

```diff
@@ -0,0 +1,29 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// Thrown to indicate that a purchase could not be finished successfully.
+///
+/// The exception should be implemented per platform. Each platform implementation
+/// should override the [shouldRetry] property and return the correct value
+/// according to the platform specific error codes.
+abstract class PurchaseException implements Exception {
+  /// Creates a [PurchaseException] with the specified error [code] and optional
+  /// [message].
+  PurchaseException({
+    required this.code,
+    this.message,
+  });
+
+  /// The error code indicating
+  final String code;
+
+  /// An human readible error message, possibly null.
+  final String? message;
+
+  /// Indicates if the action should be retried or not.
+  ///
+  /// Implementing classes should override this property and make sure the
+  /// correct value is returned based on the [code] value.
+  bool get shouldRetry;
```

Comment:
At first I was also thinking about having a common place to list all error codes and document them clearly (including if it is necessary to retry or not). But I couldn't think of a good place to do this. 

I didn't want to do it in this class (e.g. for the `errorCode` field) because that would mean we need to document platform specific error codes here for each platform and that could be tedious and difficult to read. Second we probably need to update the documentation when we add new platforms, which we'd likely forget since it will be in a different package. That is why I thought of making the class `abstract` this will force us to implement it in the platform specific layer (e.g. `AndroidPurchaseException`, `IosPurchaseException` or `WebPurchaseException`) and makes it easier to document the error codes there. 

The above inspired me to implement the `shouldRetry` property, since it will be likely that developers will catch the base `PurchaseException` class (and not the platform specific implementations) the IDE will also not show the documentation with all the specific error codes when you hover on the `errorCode` field (it will show this documentation of the base class `errorCode` field). Therefore it would be a lot easier for developers if they could check a property which indicates if they should retry.

So that was my reasoning, let me know if you agree? If not I would be happy to remove it and make sure we document all status codes in a common place).

Reply:
I still think it is a little forcing. Also, this value might make user not checking the error code carefully and blindly retrying without knowing what's the underlying error. 

And I think even `PurchaseException` doesn't need to be in the platform interface, as the platform interface is not using the class at all. We can move `PurchaseException` to InAppPurchaseConnection and have the error code documented in `InAppPurchaseConnection`. We can name it `InAppPurchaseException`

Reply:
You are right we don't need it in the platform interface so let's not worry about it here ;). 

I will remove it.

Reply:
Removed